provider/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,429 @@
+	Namespaces             Namespaces       `description:"Kubernetes namespaces" export:"true"`
+	LabelSelector          string           `description:"Kubernetes Ingress label selector to use" export:"true"`
+	IngressClass           string           `description:"Value of kubernetes.io/ingress.class annotation to watch for" export:"true"`
+	IngressEndpoint        *IngressEndpoint `description:"Kubernetes Ingress Endpoint"`
+	lastConfiguration      safe.Safe
+}
+
+func (p *Provider) newK8sClient(ctx context.Context, ingressLabelSelector string) (Client, error) {
+	ingLabelSel, err := labels.Parse(ingressLabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid ingress label selector: %q", ingressLabelSelector)
+	}
+	log.FromContext(ctx).Infof("ingress label selector is: %q", ingLabelSel)
+
+	withEndpoint := ""
+	if p.Endpoint != "" {
+		withEndpoint = fmt.Sprintf(" with endpoint %v", p.Endpoint)
+	}
+
+	var cl *clientImpl
+	if os.Getenv("KUBERNETES_SERVICE_HOST") != "" && os.Getenv("KUBERNETES_SERVICE_PORT") != "" {
+		log.FromContext(ctx).Infof("Creating in-cluster Provider client%s", withEndpoint)
+		cl, err = newInClusterClient(p.Endpoint)
+	} else {
+		log.FromContext(ctx).Infof("Creating cluster-external Provider client%s", withEndpoint)
+		cl, err = newExternalClusterClient(p.Endpoint, p.Token, p.CertAuthFilePath)
+	}
+
+	if err == nil {
+		cl.ingressLabelSelector = ingLabelSel
+	}
+
+	return cl, err
+}
+
+// Init the provider
+func (p *Provider) Init() error {
+	return p.BaseProvider.Init()
+}
+
+// Provide allows the k8s provider to provide configurations to traefik
+// using the given configuration channel.
+func (p *Provider) Provide(configurationChan chan<- config.Message, pool *safe.Pool) error {
+	ctxLog := log.With(context.Background(), log.Str(log.ProviderName, "docker"))
+	logger := log.FromContext(ctxLog)
+	// Tell glog (used by client-go) to log into STDERR. Otherwise, we risk
+	// certain kinds of API errors getting logged into a directory not
+	// available in a `FROM scratch` Docker container, causing glog to abort
+	// hard with an exit code > 0.
+	err := flag.Set("logtostderr", "true")
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Using Ingress label selector: %q", p.LabelSelector)
+	k8sClient, err := p.newK8sClient(ctxLog, p.LabelSelector)
+	if err != nil {
+		return err
+	}
+
+	pool.Go(func(stop chan bool) {
+		operation := func() error {
+			stopWatch := make(chan struct{}, 1)
+			defer close(stopWatch)
+			eventsChan, err := k8sClient.WatchAll(p.Namespaces, stopWatch)
+			if err != nil {
+				logger.Errorf("Error watching kubernetes events: %v", err)
+				timer := time.NewTimer(1 * time.Second)
+				select {
+				case <-timer.C:
+					return err
+				case <-stop:
+					return nil
+				}
+			}
+			for {
+				select {
+				case <-stop:
+					return nil
+				case event := <-eventsChan:
```

Comment:
Due to this eventsChan configuration, we end up with https://github.com/containous/traefik/issues/3667 where ingresses can be loaded before secrets.

We should look into updating this to ensure secrets are loaded first to avoid having  a metric ton of secret errors, because they haven't loaded yet.